### PR TITLE
address vcat return inconsistency

### PR DESCRIPTION
**Pull request by cjprybol**
https://github.com/JuliaStats/NullableArrays.jl/issues/167

### CODE REVIEW

**Comment by nalimilan** [comment]
Thanks, but I'd really like to avoid having to copy so much code from Base. This is https://github.com/JuliaLang/julia/issues/2326.

In the short term, I'm not sure what's the best approach. Isn't there a way to use the machinery from Base without reimplementing the full method?

File: src/nullablevector.jl [hunk 1]
```diff
@@ -311,3 +311,80 @@ function Base.empty!(X::NullableVector)
     empty!(X.isnull)
     return X
 end
+
+function Base.promote_rule{T1,T2}(::Type{T1}, ::Type{Nullable{T2}})
```
Comment (nalimilan) [inline src/nullablevector.jl]: This is not actually needed (and would have to live in Base anyway). Indeed you should never call `promote_rule` directly, but use `promote_type` instead. This already works:
```julia
julia> promote_type(Nullable{Int64}, Int64)
Nullable{Int64}
```

**Comment by nalimilan** [comment]
I guess one partial solution is to define methods like `vcat(::AbstractArray, ::NullableArray, ::AbstractArray...)` which call `Base.type_hcat` with the correct return type. Cf. https://github.com/JuliaStats/DataArrays.jl/pull/213.

File: src/nullablevector.jl [hunk 2]
```diff
@@ -311,3 +311,80 @@ function Base.empty!(X::NullableVector)
...
+    promote_rule(Nullable{T2}, T1)
+end
+
+function Base.typed_hcat{T}(::Type{T}, A::AbstractVecOrMat...)
```
Comment (nalimilan) [inline src/nullablevector.jl]: It's not OK to override Base methods like this with signatures that only involve Base types. That will affect unrelated packages and is known as type piracy.

File: src/nullablevector.jl (hunk 1 above)
Comment (cjprybol) [reply src/nullablevector.jl]: Yes, you're right. I see now that `promote_type` has the consistency I was expecting from `promote_rule`. I'll need to learn more about those differences.

```julia
julia> promote_rule(Nullable{Int64}, Int64)
Nullable{Int64}

julia> promote_rule(Int64, Nullable{Int64})
Union{}

julia> promote_type(Nullable{Int64}, Int64)
Nullable{Int64}

julia> promote_type(Int64, Nullable{Int64})
Nullable{Int64}
```

File: src/nullablevector.jl (hunk 2 above)
Comment (cjprybol) [reply src/nullablevector.jl]: That makes perfect sense. Thanks for explaining what type piracy is

**Comment by cjprybol** [comment]
I'm not sure how to do this without overwriting the functions. I tried several combinations of extending `hcat` and `vcat` for when the second input is a `NullableArray` but the calls kept utilizing the broad definitions in base. Am I missing something?

```julia
using Gallium
using NullableArrays
import Base.hcat

julia> @enter hcat(1:2, NullableArray(3:4))
In abstractarray.jl:1001
1000
1001    hcat(A::AbstractVecOrMat...) = typed_hcat(promote_eltype(A...), A...)
1002    hcat{T}(A::AbstractVecOrMat{T}...) = typed_hcat(T, A...)
1003

About to run: (Core._apply)(Base.promote_eltype,(1:2,Nullable{Int64}[3,4]))
1|debug >

julia> # try to define a more specific function definition

julia> hcat{T<:NullableArray}(A, B::T) = typed_hcat(promote_eltype(A...), A...)
hcat (generic function with 22 methods)

julia> # still calls the same function

julia> @enter hcat(1:2, NullableArray(3:4))
In abstractarray.jl:1001
1000
1001    hcat(A::AbstractVecOrMat...) = typed_hcat(promote_eltype(A...), A...)
1002    hcat{T}(A::AbstractVecOrMat{T}...) = typed_hcat(T, A...)
1003

About to run: (Core._apply)(Base.promote_eltype,(1:2,Nullable{Int64}[3,4]))
```

**Comment by cjprybol** [comment]
I need to figure out how to handle the 3+ argument case but this is working nicely for the 2 argument cases

```julia
julia> @test typeof(vcat(1:2, NullableArray(3:4))) == NullableArrays.NullableArray{Int64,1}
Test Passed
  Expression: typeof(vcat(1:2,NullableArray(3:4))) == NullableArrays.NullableArray{Int64,1}
   Evaluated: NullableArrays.NullableArray{Int64,1} == NullableArrays.NullableArray{Int64,1}

julia> @test typeof(vcat([1 2], NullableArray([3 4]))) == NullableArrays.NullableArray{Int64,2}
Test Passed
  Expression: typeof(vcat([1 2],NullableArray([3 4]))) == NullableArrays.NullableArray{Int64,2}
   Evaluated: NullableArrays.NullableArray{Int64,2} == NullableArrays.NullableArray{Int64,2}

julia> @test typeof(hcat(1:2, NullableArray(3:4))) == NullableArrays.NullableArray{Int64,2}
Test Passed
  Expression: typeof(hcat(1:2,NullableArray(3:4))) == NullableArrays.NullableArray{Int64,2}
   Evaluated: NullableArrays.NullableArray{Int64,2} == NullableArrays.NullableArray{Int64,2}

julia> @test typeof(hcat([1 2], NullableArray([3 4]))) == NullableArrays.NullableArray{Int64,2}
Test Passed
  Expression: typeof(hcat([1 2],NullableArray([3 4]))) == NullableArrays.NullableArray{Int64,2}
   Evaluated: NullableArrays.NullableArray{Int64,2} == NullableArrays.NullableArray{Int64,2}
```

File: src/nullablevector.jl [hunk 3]
```diff
@@ -311,3 +311,14 @@ function Base.empty!(X::NullableVector)
     empty!(X.isnull)
     return X
 end
+
+notnullarray = Union{filter!(x -> !isa(x, Type{NullableArray}), subtypes(AbstractArray))...}
```
Comment (nalimilan) [inline src/nullablevector.jl]: You shouldn't need this. Just `AbstractArray` should be enough, and the most specific method will take precedence.

File: src/nullablevector.jl [hunk 4]
```diff
@@ -311,3 +311,14 @@ function Base.empty!(X::NullableVector)
...
+notnullmatrix = typeintersect(notnullarray, AbstractMatrix)
+notnullvector = typeintersect(notnullarray, AbstractVector)
+
+Base.vcat{T <: notnullarray}(A::T, B::NullableArray) = Base.vcat(NullableArray(A), B)
```
Comment (nalimilan) [inline src/nullablevector.jl]: This is wasteful since it creates a copy. Maybe we can use this trick to force creating a `NullableArray`: `vcat(NullableArray{eltype(A)}(), A, B)`.

Comment (nalimilan) [reply src/nullablevector.jl]: Also, no need for `Base.`.

File: src/operators.jl [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ## Lifted operators
 
 importall Base.Operators
-import Base: promote_op, abs, abs2, sqrt, cbrt, scalarmin, scalarmax, isless
+import Base: abs, abs2, cbrt, isless, promote_op, scalarmin, scalarmax, sqrt
```
Comment (nalimilan) [inline src/operators.jl]: Unrelated.

File: test/nullablevector.jl [hunk 6]
```diff
@@ -223,4 +223,21 @@ module TestNullableVector
     X = NullableArray(A, M)
     empty!(X)
     @test isempty(X)
+
+    @test typeof(hcat(NullableArray(1:2), 3:4)) == NullableArrays.NullableArray{Int64,2}
```
Comment (nalimilan) [inline test/nullablevector.jl]: No need for `NullableArrays.`.

File: src/nullablevector.jl [hunk 7]
```diff
@@ -1,3 +1,5 @@
+import Base: hcat, vcat
```
Comment (cjprybol) [inline src/nullablevector.jl]: This file uses `Base.` to prefix the function extensions. Should I add all of the functions to this import call? It's an unrelated change but importing these two and leaving everything else to be `Base.` is inconsistent

File: src/nullablevector.jl (hunk 3 above)
Comment (cjprybol) [reply src/nullablevector.jl]: When using`AbstractArray` there's no way to exit the `vcat -> vcat` loop and it results in StackOverflow. We either eternally prepend empty NullableArrays (`vcat(NullableVector{eltype(A)}(), A, B)`) or eternally try to convert A to a `NullableArray` (`vcat(NullableArray(A), B)`)

File: src/nullablevector.jl (hunk 4 above)
Comment (cjprybol) [reply src/nullablevector.jl]: I think this case of adding an empty NullableArray to the beginning of the `vcat/hcat` call only works on `vcat` with `Vector`s, but I've added it for that case. Every other case I tried will fail in dimension mismatch checks

File: src/operators.jl (hunk 5 above)
Comment (cjprybol) [reply src/operators.jl]: reverted

File: test/nullablevector.jl (hunk 6 above)
Comment (cjprybol) [reply test/nullablevector.jl]: removed

File: src/nullablevector.jl (hunk 3 above)
Comment (cjprybol) [reply src/nullablevector.jl]: I now realize that if I call `typed_hcat`, `typed_vcat` instead of calling `hcat/vcat` I can probably avoid the StackOverflow while using the simpler `AbstractArray` as you suggested.

**Comment by cjprybol** [comment]
Should these promote to NullableCategoricalArray?
```julia
julia> vcat(categorical([1,2,3]), NullableArray([4,5,6]))
6-element NullableArrays.NullableArray{Int64,1}:
 1
 2
 3
 4
 5
 6

julia> vcat(NullableArray([4,5,6]), categorical([1,2,3]))
6-element NullableArrays.NullableArray{Int64,1}:
 4
 5
 6
 1
 2
 3
```

File: src/nullablevector.jl [hunk 8]
```diff
@@ -311,3 +313,27 @@ function Base.empty!(X::NullableVector)
     empty!(X.isnull)
     return X
 end
+
+function vcat(A::AbstractVector, B::AbstractVector...)
```
Comment (nalimilan) [inline src/nullablevector.jl]: This is still type piracy. There needs to be `NullableArray` in the signature. Currently I don't think this can be fixed in full generality. We can include a few methods for all possible combinations up to e.g. 3 arguments, i.e:
- `::NullableArray, ::AbstractArray`
- `::AbstractArray, ::NullableArray`
- `::NullableArray, ::AbstractArray, ::AbstractArray`
- `::AbstractArray, ::NullableArray, ::AbstractArray`
- `::AbstractArray, ::AbstractArray, ::NullableArray`

File: src/nullablevector.jl (hunk 7 above)
Comment (nalimilan) [reply src/nullablevector.jl]: Yeah, better not import anything and use the `Base.` prefix when overloading methods.

**Comment by nalimilan** [comment]
> Should these promote to NullableCategoricalArray?

Yes, but that's tricky since doing so forces one package to depend on the other one, and as long as StatsModels depends on CategoricalArrays we'd better get rid of the NullableArrays dependency (https://github.com/JuliaData/CategoricalArrays.jl/pull/55). So for now I would leave this case as is.

File: src/nullablevector.jl (hunk 8 above)
Comment (cjprybol) [reply src/nullablevector.jl]: I think defining these functions would be a regression compared to just leaving the current functionality as is.
```julia
# this rule seems more intuitive
isa(array[1], NullableArray) ? NullableArray{T} : Array[Nullable{T}]

# than this rule
isa(array[1], NullableArray) ||
isa(array[2], NullableArray) ||
isa(array[3], NullableArray) ? NullableArray{T} : Array[Nullable{T}]
```

If the general functionality is type piracy then I feel the best way forward would be to either address https://github.com/JuliaLang/julia/issues/2326 directly or accept the current behavior. Thoughts?

Comment (nalimilan) [reply src/nullablevector.jl]: I really don't know. Clearly we can't reach a satisfying state without improvements in Julia. But which of the partial solutions is better, I don't know...

Comment (cjprybol) [reply src/nullablevector.jl]: Similar sentiments here. I'll implement the 2 argument cases for `hcat/vcat` and now at least we have everything written up and documented.

File: src/nullablevector.jl [hunk 9]
```diff
@@ -314,26 +314,19 @@ function Base.empty!(X::NullableVector)
     return X
 end
 
-function vcat(A::AbstractVector, B::AbstractVector...)
-    any(b -> isa(b, NullableArray), B) ?
-        typed_vcat(promote_eltype(A, B...), NullableArray(A), B...) :
-        typed_vcat(promote_eltype(A, B...), A, B...)
+function vcat(A::AbstractVector, B::NullableVector)
+        typed_vcat(promote_eltype(A, B), NullableArray(A), B)
```
Comment (ararslan) [inline src/nullablevector.jl]: Please reduce the number of spaces used for indentation here by 50%

File: src/nullablevector.jl (hunk 8 above)
Comment (nalimilan) [reply src/nullablevector.jl]: If you implement the two-argument versions, please also implement the three-argument versions. Else the rule is going to be even more complex than what you wrote above.

Of course, this rests on the assumption that concatenating more than 3 arrays is a somewhat rare operation, so that it's not a big issue if other variants behave differently. Might not be the case, but only special-casing two-argument methods is clearly a worse solution.

**Comment by cjprybol** [comment]
I think we should close this. I read through a few dozen pages of search results for `vcat` usage in Julia code here on GitHub and saw a usage pattern ~ `vcat(a,b) >> vcat(a...) ~= vcat(a,b,c) >> vcat(a,b,c,d)`. If we could redirect this into a brainstorm about how to address https://github.com/JuliaLang/julia/issues/2326, I think that would be a much better use of our time and I'd be happy to try and implement suggestions. I tried variants of https://github.com/JuliaLang/julia/issues/2326#issuecomment-44790021 and couldn't get any of them to work

**Comment by nalimilan** [comment]
I've tried starting a discussion about the way to move forward at https://github.com/JuliaLang/julia/pull/20815, let's see what can be done about this.

**Comment by ararslan** [comment]
In the meantime, thanks so much for your hard work here, Cameron. Much appreciated.
